pkg/resource/deploy/deployment_executor.go

```diff
@@ -216,6 +221,18 @@ func (ex *deploymentExecutor) Execute(callerCtx context.Context, opts Options, p
 	//     and we need to bail. This can also happen if the user hits Ctrl-C.
 	canceled, err := func() (bool, error) {
 		logging.V(4).Infof("deploymentExecutor.Execute(...): waiting for incoming events")
+		var workerPool *workerPool
+		// Use a pool of workers to run step generation in parallel if the env
+		// variable PULUMI_PARALLEL_STEP_GEN is truthy.
+		if UseParallelStepGen || env.ParallelStepGeneration.Value() {
+			if opts.InfiniteParallelism() {
+				workerPool = newWorkerPool(0, cancel)
+			} else {
+				workerPool = newWorkerPool(opts.DegreeOfParallelism(), cancel)
+			}
+			logging.V(4).Infof("deploymentExecutor.Execute(...): using %d workers", workerPool.numWorkers)
+		}
```

Comment:
can we encapsulate the logic here being the worker pool? i.e. always have a worker pool and let it make the decisions about where work happens?

Reply:
I would take this as a following change. It is the correct approach, but I wanted to make this as minimally intrusive as possible when the feature isn't enabled.